### PR TITLE
feature: added new section --- post_setup_server_root

### PR DESCRIPTION
**Pull request by leslie-tsang**
add new section `post_setup_server_root` to dynamic init server after `setup_server_root`

allow user to custom setup `nginx` runtime before launch up

### CODE REVIEW

**Comment by leslie-tsang** [comment]
@zhuizhuhaomeng Cloud you take a moment to review this PR :)

**Comment by zhuizhuhaomeng** [comment]
I don't know how to use --- post_init.
You need to add doc about post_init. 
And you need to add an example also.

File: lib/Test/Nginx/Socket.pm [hunk 1]
```diff
@@ -3283,6 +3283,20 @@ current test block. Default to "localhost".
 
 Run a piece of Perl code specified as the content of this C<--- init> section before running the tests for the blocks. Note that it is only run once before *all* the repeated requests for this test block.
 
+=head2 post_init
```
Comment (zhuizhuhaomeng) [inline lib/Test/Nginx/Socket.pm:3286]: post_init is not a good name.
I don't know what the init means.

File: lib/Test/Nginx/Util.pm [hunk 2]
```diff
@@ -915,6 +916,18 @@ sub setup_server_root ($) {
         bail_out "Failed to do mkdir $ConfDir\n";
 }
 
+sub setup_server_env ($) {
```
Comment (zhuizhuhaomeng) [inline lib/Test/Nginx/Util.pm:919]: The setup_server_env is not a good name either.

**Review by zhuizhuhaomeng — CHANGES_REQUESTED** [review]

File: lib/Test/Nginx/Socket.pm (hunk 1 above)
Comment (leslie-tsang) [reply lib/Test/Nginx/Socket.pm:3286]: IMO, `--- post_init`'s `init` inherited from `--- init`.

Both section can do the same thing, the difference is `--- post_init` is executed later than `--- init`, which allow it to preserve the changes in `t/servroot/conf`, then we can test the stuff within the `conf` folder easier.

A typical scenario is that we are able to test some of the conf files directly in a classic nginx workbench like bellow

```shell
├── conf
│   ├── mime.types
│   └── nginx.conf
└── t
    ├── firmware.pm
    └── sanity.t
```

In this scenario, we can use `--- post_init` to create a symbolic link for `conf/mime.types`, which allow test nginx instance to include it by `include mime.types;`, it do help to reduce the complexity to use `include` in test nginx instance. :)

File: lib/Test/Nginx/Util.pm (hunk 2 above)
Comment (leslie-tsang) [reply lib/Test/Nginx/Util.pm:919]: Agreed, but I haven't come up with a good function name yet, and using functions is just to minimize redundant code.

Comment (leslie-tsang) [reply lib/Test/Nginx/Util.pm:919]: Any suggestion for this? or directly copy the function code and replace it at its original location?

File: lib/Test/Nginx/Socket.pm (hunk 1 above)
Comment (zhuizhuhaomeng) [inline lib/Test/Nginx/Socket.pm:3286]: how about post_setup_server_root

File: lib/Test/Nginx/Util.pm (hunk 2 above)
Comment (zhuizhuhaomeng) [reply lib/Test/Nginx/Util.pm:919]: how about run_customer_init_script

**Review by zhuizhuhaomeng — CHANGES_REQUESTED** [review]

File: lib/Test/Nginx/Socket.pm (hunk 1 above)
Comment (leslie-tsang) [reply lib/Test/Nginx/Socket.pm]: LGTM, thanks for advice. :)

The new one makes it easier to understand without a source code background!

**Review by zhuizhuhaomeng — APPROVED** [review]

**Pull request merged**